inst/include/utils.h

```diff
@@ -1,5 +1,75 @@
 #include <torch.h>
 
+#include <stack>
+#include <mutex>
+#include <condition_variable>
+#include <future>
+#include <functional>
+#include <utility>
+#include <type_traits>
+
+template <typename F>
+class ScopeGuard {
+ public:
+  explicit ScopeGuard(F&& f) noexcept : f_(std::forward<F>(f)) {}
+  ~ScopeGuard() noexcept { f_(); }
+ private:
+  typename std::decay<F>::type f_;
+};
+
+// TODO:
+// create a factory function for ScopeGuard that disallows the returned
+// ScopeGuard object to be discarded immediately (but we need to do so in a
+// portable manner, by using the [[nodiscard]] attribute if possible, or
+// whatever its equivalent may be in other compilers that need to build {torch}
+// but don't support [[nodiscard]] yet)
+template <typename F>
+ScopeGuard<F> makeScopeGuard(F&& f) {
+  return ScopeGuard<F>(std::forward<F>(f));
+}
+
+template <typename T>
+class EventLoop {
+ public:
+  EventLoop() = default;
+  void run() {
+    while (true) {
+      std::packaged_task<T()> fn;
+      {
+        std::unique_lock<std::mutex> lock(mtx_);
+        if (tasks_.empty()) {
+          cv_.wait(lock, [this] { return !tasks_.empty(); });
```

Comment:
- Minor improvement: using condition variable to make this thread up after the task queue goes from empty to non-empt.y
- Another related minor improvement: not calling `unlock()` or `lock()` manually here (as doing so here while also ensuring exception-safety is a bit error-prone)